src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -628,205 +199,71 @@ protected override void Dispose(bool disposing)
 			}
 		}
 
-		private void DoClose()
-		{
-			if (m_command != null)
-			{
-				while (NextResult())
-				{
-				}
-
-				Reset();
-				m_session = null;
-
-				var connection = m_command.Connection;
-				connection.HasActiveReader = false;
-				m_command.ReaderClosed();
-				if (m_behavior.HasFlag(CommandBehavior.CloseConnection))
-				{
-					m_command.Dispose();
-					connection.Close();
-				}
-				m_command = null;
-			}
-		}
+		internal MySqlCommand Command { get; private set; }
+		internal MySqlConnection Connection => Command?.Connection;
+		internal MySqlSession Session => Command?.Connection.Session;
 
 		internal static async Task<MySqlDataReader> CreateAsync(MySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken)
 		{
 			var dataReader = new MySqlDataReader(command, behavior);
-			await dataReader.ReadResultSetHeaderAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+			await dataReader.ReadFirstResultSetAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
 			return dataReader;
```

Comment:
In order to buffer the entire result set, add `while (await BufferNextResultAsync(ioBehavior, cancellationToken).ConfigureAwait(false) != null){}` before the return